src/librustc_typeck/check/mod.rs

```diff
@@ -3005,6 +3045,38 @@ enum TupleArgumentsFlag {
     TupleArguments,
 }
 
+fn expected_types_for_fn_args<'a, 'tcx>(fcx: &FnCtxt<'a, 'tcx>,
+                                        expected_ret: Expectation<'tcx>,
+                                        formal_ret: ty::FnOutput<'tcx>,
+                                        formal_args: &[Ty<'tcx>])
+                                        -> Vec<Ty<'tcx>> {
+    // Unify the return type with the expected type early, for more coercions.
+    let expected_args = expected_ret.only_has_type(fcx).and_then(|ret_ty| {
+        if let ty::FnConverging(formal_ret_ty) = formal_ret {
+            fcx.infcx().probe(|_| {
+                use rustc::middle::infer::combine::Combine;
+                let trace = infer::TypeTrace::dummy(fcx.tcx());
+                match fcx.infcx().equate(false, trace).tys(ret_ty, formal_ret_ty) {
```

Comment:
I execute code in the `Ok` case, before rolling back the snapshot, while `mk_eqty` would commit it. Unless you mean it's okay to have nested snapshots like this, in which case I have to wonder what the cost is.